homeassistant/components/weather/__init__.py

```diff
@@ -322,6 +326,55 @@ def __init_subclass__(cls, **kwargs: Any) -> None:
                     cls.__name__,
                     report_issue,
                 )
+        if any(method in cls.__dict__ for method in ("_attr_forecast", "forecast")):
+            if _reported_forecast is False:
+                module = inspect.getmodule(cls)
+                _reported_forecast = True
+                if (
+                    module
+                    and module.__file__
+                    and "custom_components" in module.__file__
+                ):
+                    report_issue = "report it to the custom integration author."
+                else:
+                    report_issue = (
+                        "create a bug report at "
+                        "https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue"
+                    )
+                _LOGGER.warning(
+                    (
+                        "%s::%s is using a forecast attribute on an instance of "
+                        "WeatherEntity, this is deprecated and will be unsupported "
+                        "from Home Assistant 2024.3. Please %s"
+                    ),
+                    cls.__module__,
+                    cls.__name__,
+                    report_issue,
+                )
+
+    @callback
+    def add_to_platform_start(
+        self,
+        hass: HomeAssistant,
+        platform: EntityPlatform,
+        parallel_updates: asyncio.Semaphore | None,
+    ) -> None:
+        """Start adding an entity to a platform."""
+        super().add_to_platform_start(hass, platform, parallel_updates)
+        ir.async_create_issue(
```

Comment:
As I understand it, this unconditionally creates an issue regardless of if the weather entity implements legacy forecast or not. To fix that, maybe set a flag on the class in `__init_subclass__` and then check that flag here?

For example:
```py
def __init_subclass__(cls, **kwargs: Any) -> None:
    if any(method in cls.__dict__ for method in ("_attr_forecast", "forecast")):
        cls.__weather_legacy_forecast = True

    @callback
    def add_to_platform_start(
        self,
        hass: HomeAssistant,
        platform: EntityPlatform,
        parallel_updates: asyncio.Semaphore | None,
    ) -> None:
        """Start adding an entity to a platform."""
        super().add_to_platform_start(hass, platform, parallel_updates)
        if self.__weather_legacy_forecast and not self.__weather_legacy_forecast_reported:
            ir.async_create_issue(
```